### ISSUE TITLE
IFrame elements being filtered by host domains instead of src

### ISSUE THREAD

**Issue by Sorrow-San** [AdguardTeam/AdguardForiOS#1393]
### Core issue:

Much like the title suggests, I've noticed that the AdGuard IOS application is inconsistent with iFrame / Subdocument elements when adding user-rules. Instead of registering the elements within a iFrame (both cosmetic & domain request origins) to the src domain of the iFrame it often registers the elements to the host domain instead.

----

### Example:
If you load onto **websiteA** which has an iFrame element with a src of **websiteB**, any rules that target **websiteB** such as **websiteB##.advertisement** or ***$domain=websiteB,third-party** won't properly work. However, if those same rules are created for **websiteA** then they'll apply to any iFrame with any src on **websiteA**.

-----

### Additional information:

This isn't 100% consistent and appears to be somehow related to when the iFrame loads, although this is difficult to prove without having timings. However, there has been times where filters in the above example properly apply then if the iFrame were to have the src changed or be re-created then those filters would abruptly stop working again.

-----

### Why this is an issue:

This is a big deal when it comes to websites that have multiple iFrame elements from different sources on them, because you're forced to apply the filter to the host domain in order to consistently affect the iFrames. This means a filter you only need for one iFrame will also be applied to all other frames, possibly even breaking them.

**Comment by SoLenny** [comment]
@Sorrow-San Hello. Could you give real examples using the rules? @Alex-302 will have a look.

**Comment by Sorrow-San** [comment]
> 
> 
> @Sorrow-San Hello. Could you give real examples using the rules? @Alex-302 will have a look.

Hello SoLenny, the following is one real example using real rules and a real website.

------

### Rules

Blocking all domains which originate from prettyfast.to
*$domain=prettyfast.to

"Whitelisting" needed domains for iFrame to function.
@@||jwpsrv.com
@@||googleapis.com
@@||googleusercontent.com
@@||gstatic.com

------

### Example of site with rules in effect 

https://9anime.to/watch/phantasy-star-online-2-episode-oracle.k5lr/r2z00o

When loading the **F5 server**, (assuming it's not down again to be worked on when you go to check) the src of the iFrame will be set to **prettyfast.to**, which should then bring the first rule into effect. 

**Note**: Prettyfast.to is closer to a CDN that holds the javascript needed for a jwplayer based video player to load and pull down video data.


This rule should prevent doubooca.com, psitchie.com, serconmp.com, and some other advertising domains from loading within the iFrame, however, it doesn't affect them at all. Even though they definitely are loading off Prettyfast.to inside said iFrame.

**Second Note**: There should be no problem with the rules themselves as the same method can be used on desktop implementations of multiple adBlockers and work without issue.

------

### Why it matters in this case

The previously mentioned advertising domains which are loaded inside the iFrame tend to change name once every day or so. Since (as far as I'm aware) regex expressions and some other techniques that could be used instead in this situation aren't supported by the **IOS AdGuard application**, I have to reverse my method of dealing with the issue by blocking every third-party url that originates from prettyfast.to first then unblocking the individual urls which are needed for the iFrame to function.

------

I'll be away for a few days due to the holidays, however, when I return I'll answer any other questions about the issue you may have.

Thank you,
and a very happy Holiday to everyone!

**Comment by Alex-302** [comment]
> Blocking all domains which originate from prettyfast.to
*$domain=prettyfast.to

this rule blocks any request at prettyfast.to

> "Whitelisting" needed domains for iFrame to function.

In Safari try to use domain of the site instead on iframe's.

**Comment by Sorrow-San** [comment]
@Alex-302 Please read the original issue in question instead of picking apart my example.


The issue is...
I Frames normally register AS the domain they contain, for rules. NOT the domain that created the iFrame.

IE: Currently rules need to be made like this *$domain=9Anime.to to block advertisements that Originate within the iFrame which has a SRC of Prettyfast.to

This is **BAD** because it’ll affect everything on 9Anime.to, including any IFRAMEs created on the site REGARDLESS of the src they use.

And it’s also **WRONG**, because the Origin of the advertisements were the Prettyfast.to IFrame and not 9Anime.to even though the IFrame containing prettyfast.to originates from 9Anime.

----

Here, maybe this will help.

1. Filters used with AdblockPlus on IOS
https://gitlab.com/Sorrow-San/9anime-adblock-filters/raw/AdGuard_filter_list/9Anime%20filter%20list

Notice that the first rule HAS to include the 9Anime domains? Even though Prettyfast.to and other domains are in there they actually do nothing because the iframes are only affected by filters for the site the iFrame is on. In this case, 9Anime


2. Same filters for literally anything else
https://gitlab.com/Sorrow-San/9anime-adblock-filters/raw/master_EasyListFormat/9Anime%20filter%20list.txt

Instead of needing to block 9Anime with the rules. They properly understand that a rule such as *$domain=prettyfast.to should affect an iFrame using this as the SRC.

**Comment by ameshkov** [comment]
> Unless you’re telling me that this is a problem with Safari recording domain origins wrong.

That's exactly what he's telling unfortunately

**Comment by ameshkov** [comment]
Moreover, I reported it to Apple four years ago and asked to change this.

There is an option to use `if-top-url` condition in Safari content blocking that could help with this. But there's another issue with it: you cannot mix `if-top-url` and `if-domain` in a single content blocker file, and replacing all `if-domain` with `if-top-url` will make it insanely slow.

**Comment by Sorrow-San** [comment]
@ameshkov Thank you for clarifying, i guess the hacky setup I’m using will have to suffice for the time being.

Apologies for the trouble. And I appreciate the tip.

**Comment by krystian3w** [comment]
https://github.com/FiltersHeroes/PolishAnnoyanceFilters/issues/2974
